File: LeetCode/Medium/5 - Longest Palindromic Substring/5 - Longest Palindromic Substring.js
/**
 * @param {string} s
 * @return {string}
 */
var longestPalindrome = function(s) {
    const n = s.length;
    if (!n) return '';
    
    const getLen = (l, r) => {
        while (l >= 0 && r < n && s[l] == s[r]) {
            l--;
            r++;
        }
        return r - l -1;
    }
    
    let l = 0, len = 0;
    for (let i = 0; i < n; ++i) {
        const curLen = Math.max(getLen(i, i), getLen(i, i + 1));
        
        if (curLen > len) {
            l = i - ((curLen - 1) >> 1);
            len = curLen;
        }
    }
    return s.substr(l ,len);
};

File: LeetCode/Medium/5 - Longest Palindromic Substring/5 - Longest Palindromic Substring.cpp
class Solution {
public:
    int n;
    
    string longestPalindrome(string s) {
        n = s.size();
        int st = 0, len = 0;
        for (int i = 0; i < n; ++i) {
            int len1 = solve(i, i, s);
            int len2 = solve(i, i + 1, s);
            int mx = max(len1, len2);
            
            if (len < mx) {
                st = i - (mx - 1) / 2;
                len = mx;
            }
        }
        return s.substr(st, len);
    }
    
    int solve(int lo, int hi, string &s) {
        while (lo >= 0 && hi < n && s[lo] == s[hi]) {
            lo--;
            hi++;
        }
        return hi - lo - 1;
    }
};